modules/Auth/lib/ReservationHandler.cpp

```diff
@@ -1,110 +1,805 @@
+            return types::reservation::ReservationResult::Rejected;
+        }
+
+        const types::evse_manager::ConnectorTypeEnum connector_type =
+            reservation.connector_type.value_or(types::evse_manager::ConnectorTypeEnum::Unknown);
+        if (!is_reservation_possible(connector_type, this->global_reservations, this->evse_reservations)) {
+            print_reservations_debug_info(reservation, evse_id, true);
+            return get_reservation_evse_connector_state(connector_type);
+        }
+
+        EVLOG_info << "Created reservation for connector type "
+                   << types::evse_manager::connector_type_enum_to_string(
+                          reservation.connector_type.value_or(types::evse_manager::ConnectorTypeEnum::Unknown));
+        global_reservations.push_back(reservation);
+        store_reservations();
     }
 
-    if (date::utc_clock::now() > Everest::Date::from_rfc3339(reservation.expiry_time)) {
-        EVLOG_debug << "Rejecting reservation because expiry_time is in the past";
+    set_reservation_timer(reservation, evse_id);
+
+    return types::reservation::ReservationResult::Accepted;
+}
+
+void ReservationHandler::on_connector_state_changed(const ConnectorState connector_state, const uint32_t evse_id,
+                                                    const uint32_t connector_id) {
+    if (connector_state == ConnectorState::AVAILABLE) {
+        // Nothing to cancel.
+        return;
+    }
+
+    std::unique_lock<std::recursive_mutex> lock(reservation_mutex);
+    std::unique_lock<std::recursive_mutex> lock2(evse_mutex);
+    if (this->evses.count(static_cast<int32_t>(evse_id)) == 0) {
+        EVLOG_warning << "on_connector_state_changed: evse " << evse_id << " does not exist. This should not happen.";
+        return;
+    }
+
+    auto& connectors = this->evses[evse_id]->connectors;
+    auto connector_it = std::find_if(connectors.begin(), connectors.end(),
+                                     [connector_id](const auto& connector) { return connector_id == connector.id; });
+
+    if (connector_it == connectors.end()) {
+        // Connector with specific connector id not found
+        EVLOG_warning << "Could not change connector state for connector id " << connector_id << " of evse " << evse_id
+                      << ": connector id does not exist. This should not happen.";
+        return;
+    }
+
+    if (evse_reservations.count(evse_id) != 0 && evse_reservations[evse_id].connector_type.has_value() &&
+        (connector_it->type == evse_reservations[evse_id].connector_type.value() ||
+         connector_it->type == types::evse_manager::ConnectorTypeEnum::Unknown ||
+         evse_reservations[evse_id].connector_type.value() == types::evse_manager::ConnectorTypeEnum::Unknown)) {
+        cancel_reservation(evse_reservations[evse_id].reservation_id, true,
+                           types::reservation::ReservationEndReason::Cancelled);
+        return;
+    }
```

Comment:
nit: I find that quite hard to read. Id prefer something like:

```cpp
bool reservation_exists = evse_reservations.count(evse_id) != 0;
bool reserved_connector_specified = evse_reservations[evse_id].connector_type.has_value();
bool connector_type_matches = 
    connector_it->type == evse_reservations[evse_id].connector_type.value();
bool connector_type_unknown =
    connector_it->type == types::evse_manager::ConnectorTypeEnum::Unknown ||
    evse_reservations[evse_id].connector_type.value() == types::evse_manager::ConnectorTypeEnum::Unknown;

if (reservation_exists && reserved_connector_specified && (connector_type_matches || connector_type_unknown)) {
    cancel_reservation(evse_reservations[evse_id].reservation_id, 
                       true, 
                       types::reservation::ReservationEndReason::Cancelled);
    return;
}
```

Reply:
Hmm I did this, but the order is important because of the optionals. I tried what happens if I keep those bools but it's a lot extra checks and I don't know if it will be more readable.